rocketpy/environment/fetchers.py

```diff
@@ -0,0 +1,433 @@
+third-party APIs. As this is a recent module (introduced in v1.2.0), some
+functions may be changed without notice in future feature releases.
+"""
+
+import re
+import time
+from datetime import datetime, timedelta, timezone
+
+import netCDF4
+import requests
+
+from rocketpy.tools import exponential_backoff
+
+
+@exponential_backoff(max_attempts=3, base_delay=1, max_delay=60)
+def fetch_open_elevation(lat, lon):
+    """Fetches elevation data from the Open-Elevation API at a given latitude
+    and longitude.
+
+    Parameters
+    ----------
+    lat : float
+        The latitude of the location.
+    lon : float
+        The longitude of the location.
+
+    Returns
+    -------
+    float
+        The elevation at the given latitude and longitude in meters.
+
+    Raises
+    ------
+    RuntimeError
+        If there is a problem reaching the Open-Elevation API servers.
+    """
+    print(f"Fetching elevation from open-elevation.com for lat={lat}, lon={lon}...")
+    request_url = (
+        "https://api.open-elevation.com/api/v1/lookup?locations" f"={lat},{lon}"
+    )
+    try:
+        response = requests.get(request_url)
+    except requests.exceptions.RequestException as e:
+        raise RuntimeError("Unable to reach Open-Elevation API servers.") from e
+    results = response.json()["results"]
+    return results[0]["elevation"]
+
+
+@exponential_backoff(max_attempts=5, base_delay=2, max_delay=60)
+def fetch_atmospheric_data_from_windy(lat, lon, model):
+    """Fetches atmospheric data from Windy.com API for a given latitude and
+    longitude, using a specific model.
+
+    Parameters
+    ----------
+    lat : float
+        The latitude of the location.
+    lon : float
+        The longitude of the location.
+    model : str
+        The atmospheric model to use. Options are: ecmwf, GFS, ICON or ICONEU.
+
+    Returns
+    -------
+    dict
+        A dictionary containing the atmospheric data retrieved from the API.
+
+    Raises
+    ------
+    ValueError
+        If an invalid response is received from the API.
+    """
+    model = model.lower()
+    if model[-1] == "u":  # case iconEu
+        model = "".join([model[:4], model[4].upper(), model[5:]])
+
+    url = (
+        f"https://node.windy.com/forecast/meteogram/{model}/{lat}/{lon}/"
+        "?step=undefined"
+    )
+
+    try:
+        response = requests.get(url).json()
+        if "data" not in response.keys():
+            raise ValueError(
+                f"Could not get a valid response for {model} from Windy. "
+                "Check if the coordinates are set inside the model's domain."
+            )
+    except requests.exceptions.RequestException as e:
+        if model == "iconEu":
+            raise ValueError(
+                "Could not get a valid response for Icon-EU from Windy. "
+                "Check if the coordinates are set inside Europe."
+            ) from e
+
+    return response
```

Comment:
When the user incorrectly specifies the `model`, i.e. it is not a valid model name , the fetcher runs for a whole 1 minute and then returns a `UnboundLocalError` exception. Maybe we should check beforehand if the model name is correct and raise a more informative exception.

Below is a code to reproduce the error. Using "ECMWF" for `file` runs correctly, but with the incorrect model name it gives the incorrect behavior described above. Note that the `file` argument for `set_atmospheric_model` is eventually passed as the `model` argument for `fetch_atmospheric_data_from_windy`.

```
EnvWindy = Environment(
    date=(2019, 2, 10, 18),
    latitude=-21.960641,
    longitude=-47.482122,
)
EnvWindy.set_atmospheric_model(
    type="Windy",
    file="oh_no_wrong_model_name",
    #file="ECMWF",
)
```

Reply:
Hey, this is a nice observation.
I added a validation to the Environemnt class, before calling this fetcher function.

I think it is not a good idea to raise an exception inside the fetcher since it would still run indefinitelly until the max_attemp count reaches the limit.